internal/archive/credentials.go

```diff
@@ -0,0 +1,297 @@
+package archive
+
+import (
+	"bufio"
+	"errors"
+	"fmt"
+	"io"
+	"io/fs"
+	"net/url"
+	"os"
+	"path/filepath"
+	"sort"
+	"strings"
+)
+
+// This file defines findCredentials() function for searching repository
+// credentials in Apt configuration, see
+// https://manpages.debian.org/testing/apt/apt_auth.conf.5.en.html.
+
+// credentials represents matched Username and Password if Username is
+// non-empty or unsuccessful search otherwise.
+type credentials struct {
+	Username string
+	Password string
+}
+
+// Empty checks whether c represents unsuccessful search.
+func (c credentials) Empty() bool {
+	return c.Username == ""
+}
+
+// credentialsLookup contains parsed input URL data used for search.
+type credentialsLookup struct {
+	scheme     string
+	host       string
+	port       string
+	path       string
+	needScheme bool
+}
+
+// lookupFor parses repoUrl into credentialsLookup and fills provided credentials with
+// username and password if they are specified in repoUrl.
+func lookupFor(repoUrl string, creds *credentials) (*credentialsLookup, error) {
```

Comment:
I'm ambivalent about this one. You are right it's a _return_ parameter. But `credentials` structure is defined in return values of `findCredsInDir` (used to be called `findCredentialsDir`) as a value (not a pointer). So we'd have to change the return value to pointer. But then, before `queryFor` (used to be `lookupFor`) is called, `creds` would be nil. So when `credsDir` doesn't exist caller would use `nil` as indicator that no credentials were found. But if it does exists, we'd have to add
```
if creds.Username == "" {
  return nil
}
```
to `findCredsInDir` so that we don't have two indicators of _credentials not found_ (`nil` and empty username).